### PR TITLE
Introduced hp::Refinement namespace.

### PR DESCRIPTION
**Pull request by marcfehling**
As pointed out in step-27 for hp adaptive methods, we need decision criteria whether h or p adaptation should be supplied. The tutorial presents one particular decision algorithm and implements it from scratch., but we could provide a selection of different algorithms to the use

This PR introduces a namespace with a selection of different decision algorithms. We tried to provide this functionality as general as possible, so that it could be easily adopted to individual needs. Please see the namespace documentation for a detailed description.

Part of #3511 and #7515.

@bangerth @davydden -- FYI

### CODE REVIEW

**Comment by marcfehling** [comment]
Many of the provided implementations contain similar syntax.

It would be worth a thought to reduce duplicates: Each function `p_adaptivity_from_threshold()`, `p_adaptivity_from_regularity()` and `p_adaptivity_from_prediction` could prepare a container of flags itself and forward it to the function `p_adaptivity_from_flags()`. This would add another loop over all cells, but would make the code easier to read.

I'll prepare a corresponding commit. You may decide which path to take ;)

**Comment by marcfehling** [comment]
Namespace and function names are open for discussion. :)

Forgot to mention that the error prediction algorithm by Melenk corresponding to `p_adaptivity_from_prediction()` will be provided in a follow-up PR.

**Comment by marcfehling** [comment]
ping

File: include/deal.II/hp/refinement.h [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+}
+
+namespace hp
+{
+  /**
+   * We supply adaptive methods to align computational ressources with the
+   * complexity of the numerical solution. Error estimates are an appropriate
+   * means of determining where adjustments need to be made.
+   *
+   * However with hp adaptivity, we have two ways to realise these adjustments:
```
Comment (kronbichler) [inline include/deal.II/hp/refinement.h]:
```suggestion
   * However with hp adaptivity, we have two ways to realize these adjustments:
```
(We try to consistently use American English spelling in deal.II.)

File: include/deal.II/hp/refinement.h [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+   * For irregular solutions, h adaptive methods which dynamically assign cell
+   * sizes tend to reduce the approximation error, while for smooth solutions p
+   * adaptive methods are better suited in which function spaces will be
+   * selected dynamically. This namespace offers all tools to decide which type
```
Comment (kronbichler) [inline include/deal.II/hp/refinement.h]:
```suggestion
   * selected dynamically. This namespace collects tools to decide which type
```

File: include/deal.II/hp/refinement.h [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+      const double                         p_refine_fraction  = 0.5,
+      const double                         p_coarsen_fraction = 0.5);
+
+    /**
+     * Adapt the finite element on cells based on the regularity of the
+     * (unknown) analytical solution.
+     *
+     * With an approximation of the local Sobolev regularity index $k_K$,
+     * we may assess to which finite element space our local solution on cell
+     * $K$ belongs. Since the regularity index is only an etimate, we won't
```
Comment (kronbichler) [inline include/deal.II/hp/refinement.h]:
```suggestion
     * $K$ belongs. Since the regularity index is only an estimate, we won't
```

**Pull request merged**